proposals/p2240.md

```diff
@@ -0,0 +1,758 @@
+programmers who have less familiarity with variadics.
+
+This decision was made in
+[leads issue #1162](https://github.com/carbon-language/carbon-lang/issues/1162).
+
+### Use postfix instead of prefix operators
+
+`...` is a postfix operator in C++, and `[:]` is a postfix operator in C++
+proposal P1858R2. Furthermore, in both cases postfix position is arguably more
+natural: postfix `...` aligns with the natural-language usage of "…", and
+postfix `[:]` aligns with C++'s (and Carbon's) subscripting syntax. So in some
+ways it would be very desirable to make their Carbon counterparts postfix as
+well. Notably, it would let us describe variadic evaluation very cleanly in
+terms of replacing `[:]` with `[$I]`. However, this turns out to be an awkward
+fit for Carbon. Consider this example:
+
+```
+fn SumInts(..., [:]params: [i64;]) -> i64
+```
+
+If `[:]` and `...,` were suffix operators, this would instead be:
+
+```
+fn SumInts(params: [i64;][:] ...,) -> i64
+```
+
+But this is ambiguous about whether the `[:]` operator applies to
+`params: [i64;]` or just `[i64;]`, so we need to add a set of parentheses:
+
+```
+fn SumInts((params: [i64;])[:] ...,) -> i64
+```
+
+To state the point more generally: when an expression mixes prefix and suffix
+operators, their order is determined by precedence and parentheses, which is
+usually less readable than relying on the linear order of operators with the
+same fixity. The binding operator `:` is effectively a prefix operator (because
+its left-hand side is a single token), so if `[:]` is a suffix operator, code
+that mixes the two will be somewhat harder to read. That's a particular problem
+because, as our motivating examples illustrate, variadic function signatures
+will often have those two operators in close conjunction, and readability is at
+a premium in function signatures.
+
+C solved a very similar problem with mixing prefix `*` and postfix `.` by
+introducing a separate `->` operator. We could in principle solve this issue the
+same way, introducing a separate "variadic binding operator" `$`, such that
+`params$ [i64;]` is equivalent to `(params: [i64;])[:]`. However, this would
+trade a readability problem for a teachability problem, by introducing more
+syntax that programmers need to learn before they can even read variadic
+function signatures.
+
+### Keyword syntax
+
+> **TODO:** Decide whether to make this the primary proposal, and restructure
+> the discussion as appropriate.
+
+We could build the syntax around keywords rather than symbolic tokens, much like
+the
+[Swift proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md).
+For concreteness, the keywords I will discuss here are as follows:
+
+| Symbolic token | Keyword replacement               |
+| -------------- | --------------------------------- |
+| `[:]`          | `expand`                          |
+| `...,`         | `repeat` and `unpack` (see below) |
```

Comment:
Have you considered using `...` instead of `...,`? While the consistency between `...,` and the other `...blah` tokens is nice, I think the presence of a `,` in the `...,` token is going to be a persistent readability issue, because it makes it visually unclear whether you're expanding the thing to the left or the thing to the right.

Eg:

```
fn F(...each x: i32) {
  if (...and each x > 0) {
    ...{
      Print(each x);
    }
    F(...each x - 1);
  }
}
```

In in-person discussion between @chandlerc, @josh11b and myself, there was some sentiment towards using prefix `...` in patterns and postfix `...` in expressions, but I'm not sure whether that will work well in the overall grammar, particularly when expressions appear within patterns.

I'm going to add some suggestions below using this syntax to make it easier to compare against other options.

Reply:
Mm. I think reading `...each` as a folding operator actually works well in the pattern case, where there can't be any postfix pieces following the `each ...` subexpression, but yeah, in the `F(...each x - 1)` case it's easy to misread it in a way where `F(...1 - each x)` would be surprising.